src/bootstrap/tool.rs

```diff
@@ -512,8 +512,10 @@ impl Step for Rustdoc {
         // Similar to `compile::Assemble`, build with the previous stage's compiler. Otherwise
         // we'd have stageN/bin/rustc and stageN/bin/rustdoc be effectively different stage
         // compilers, which isn't what we want. Rustdoc should be linked in the same way as the
-        // rustc compiler it's paired with, so it must be built with the previous stage compiler.
-        let build_compiler = builder.compiler(target_compiler.stage - 1, builder.config.build);
+        // rustc compiler it's paired with, so it must be built with the previous stage compiler,
+        // if download_rustc is true, we will use the same target stage.
+        let target_stage = target_compiler.stage - if builder.download_rustc() { 0 } else { 1 };
+        let build_compiler = builder.compiler(target_stage, builder.config.build);
```

Comment:
I'm trying to recall why we decided on this logic. I remember you telling me that we get a panic without this change, because we try and build rustdoc with stage 1, and it was deleted by your new code. But I think it would be fine to just not delete stage 1, and have that work for `rustdoc +stage1`?

OH, I just remembered why - `rustdoc +stage1` will be the compiler downloaded from nightly, even though we expect it to be the rustdoc newly built from in-tree sources. So we want to delete stage1/ so we don't use that rustdoc by accident.

This all seems quite confusing. I wonder if we should add a new `ci-rustc-sysroot/` that avoids all these issues by never putting files in the `stage1/` sysroot at all. I think that would be fairly simple to do by modifying the `sysroot` variable in compile.rs:1118. Can you please make that change rather than modifying tool.rs?

Reply:
Yes, this seems more consistent.

Reply:
Updated the code.